dashboard/config/custom_session_store/instrumented_cookie_store.rb

```diff
@@ -0,0 +1,45 @@
+require 'action_dispatch/middleware/session/cookie_store'
+require 'objspace'
+
+module ActionDispatch
+  module Session
+    # Simple wrapper around the standard Rails decentralized cookie-based
+    # session store that we currently use in production. The primary goal is to
+    # collect some data about current usage patterns, to inform our decisions
+    # about a possible switch to a centralized session store.
+    #
+    # TODO infra: remove this helper after we've gathered enough data.
+    class InstrumentedCookieStore < CookieStore
+      METRICS_NAMESPACE = 'code-dot-org/ActionDispatch'.freeze
+      METRICS_LOG_WINDOW = 1.minute
+
+      # @see https://github.com/rails/rails/blob/ac87f58207cff18880593263be9d83456aa3a2ef/actionpack/lib/action_dispatch/middleware/session/cookie_store.rb#L104-L107
+      private def write_session(req, session_id, session_data, options)
+        log_session_info(session_id, session_data)
```

Comment:
I don't understand enough about how the whole flow here works to evaluate this tradeoff.... but since `log_session_info` is coming before `super()`, what would happen if godsforbid some unknown unknowns type exceptions starts triggering in log_session_info? Obviously eating a silent failure to flush the logs (and lower the server RAM) would be not great too, but at least we can reason pretty clearly about those impacts (RAM gets full, OOM goes to town, new instance is maybe scaled up?), and we can watch the RAM.

On the flipside, if something interfered with durability of our session cookies, even in obscure conditions, it could be mildly catastrophic, and relatively hard to predict the behavior of.

My instinct would be to want log_session_info wrapped in a catch-all rescue block, so the real write_session is basically guaranteed to be called as usual.

Reply:
See below for maybe a better variation on this idea: include a rescue that covers all of log_session_info, and clear @session_info in the rescue to avoid RAM filling up too.

Reply:
This is a great suggestion, thank you!

I also added some honeybadger logging to the error handler. This does technically introduce the possibility that an error within our honeybadger logging could bubble up to affect `write_session`, but honeybadger logging in an error handler is a pattern we use frequently in the codebase and it hasn't caused us problems yet so I *think* it should be fine.